contracts/contracts/ValidatorAuction.sol

```diff
@@ -135,7 +142,13 @@ contract ValidatorAuction is Ownable {
 
     function withdraw() public {
         require(auctionState == AuctionState.Ended || auctionState == AuctionState.Failed, "You cannot withdraw before the auction is ended or it failed.");
-        require(bids[msg.sender] > closingPrice, "The sender has nothing to withdraw.");
+        require(bids[msg.sender] > lastBidPrice, "The sender has nothing to withdraw.");
```

Comment:
This contradicts the later code There you compare it to closingPrice, which can be different. 
Now the last bidder can not get his money back if the auction fails. Please add a test for this case. 

To fix this I think we should rewrite the withdraw method to make it more expicit and therefore easier to understand.

what about:
```
assert(bid>= lastBidPrice);
if (auctionstate == Ended) {
    valueToWithdraw = bid - lastBidPrice;
} else if (auctionState == Failed) {
    valueToWithdraw = bid;
}
```